addons/sourcemod/scripting/szf/pickupweapons.sp

```diff
@@ -170,20 +170,54 @@ public Action Event_ResetPickup(Event event, const char[] name, bool dontBroadca
 void SetUniqueWeapon(int iEntity, ArrayList &aWeapons, WeaponRarity iWepRarity)
 {
 	Weapon wep;
+	TFClassType nClassFilter = GetWeaponClassFilter(iEntity);
 	
 	//If array is empty, fill it again with the weapon list
 	if (aWeapons.Length == 0)
 	{
 		delete aWeapons;
 		aWeapons = GetAllWeaponsWithRarity(iWepRarity);
 	}
-	
-	int iRandom = GetRandomInt(0, aWeapons.Length - 1);
-	
-	aWeapons.GetArray(iRandom, wep);
-	SetWeaponModel(iEntity, wep);
-	//This weapon is no longer in the pool
-	aWeapons.Erase(iRandom);
+	//filter if class specific
+	if (nClassFilter == TFClass_Unknown)
+	{
+		int iRandom = GetRandomInt(0, aWeapons.Length - 1);
+		
+		aWeapons.GetArray(iRandom, wep);
+		SetWeaponModel(iEntity, wep);
+		//This weapon is no longer in the pool
+		aWeapons.Erase(iRandom);
+		
+		return;
+	}
+	else
+	{
+		ArrayList aLocalList = new ArrayList(sizeof(Weapon));
```

Comment:
Might be better to instead of creating new array `aLocalList`, Reuse `aWeapons` by sorting array randomly, then loop through each Weapon until valid class is found.
```aWeapons.Sort(Sort_Random, Sort_Integer);```